src/main/java/frc/robot/subsystems/LocalizationCamera.java

```diff
@@ -0,0 +1,259 @@
+        SmartDashboard.putNumber("vision/" + m_cameraName + "/Target Y Distance", targetTranslation2d.getY());
+
+        SmartDashboard.putBoolean("vision/" + m_cameraName + "/isEstPoseJumpy", isEstPoseJumpy());
+        SmartDashboard.putNumberArray("vision/" + m_cameraName + "/standardDeviations", curStdDevs.getData());
+    }
+
+  private void updateEstimationStdDevs(Optional<EstimatedRobotPose> estimatedPose, List<PhotonTrackedTarget> targets) {
+    if (estimatedPose.isEmpty()) {
+      // No pose input. Default to single-tag std devs
+      curStdDevs = VisionConstants.kSingleTagStdDevs;
+      SmartDashboard.putNumber("vision/" + m_cameraName + "/standardDeviation-average-distance", Double.MAX_VALUE);
+      SmartDashboard.putString("vision/" + m_cameraName + "/standardDeviation-state", "empty");
+    } else {
+      // Pose present. Start running Heuristic
+      int numTags = 0;
+      double avgDist = 0;
+
+      // Precalculation - see how many tags we found, and calculate an
+      // average-distance metric
+      for (var tgt : targets) {
+        var tagPose = poseEstimator.getFieldTags().getTagPose(tgt.getFiducialId());
+        if (tagPose.isEmpty())
+          continue;
+        numTags++;
+        avgDist += tagPose
+            .get()
+            .toPose2d()
+            .getTranslation()
+            .getDistance(estimatedPose.get().estimatedPose.toPose2d().getTranslation());
+      }
+
+      if (numTags == 0) {
+        // No tags visible. Default to single-tag std devs
+        curStdDevs = VisionConstants.kSingleTagStdDevs;
+        SmartDashboard.putString("vision/" + m_cameraName + "/standardDeviation-state", "no tags visible");
+      } else if (numTags == 1 && avgDist > VisionConstants.VISION_DISTANCE_DISCARD) {
+        curStdDevs = VecBuilder.fill(Double.MAX_VALUE, Double.MAX_VALUE, Double.MAX_VALUE);
+        SmartDashboard.putString("vision/" + m_cameraName + "/standardDeviation-state", "target too far");
+      } else {
+        var unscaledStdDevs = numTags > 1 ? VisionConstants.kMultiTagStdDevs : VisionConstants.kSingleTagStdDevs;
+
+        avgDist /= numTags;
+        // increase std devs based on (average) distance
+        curStdDevs = unscaledStdDevs.times(1 + (avgDist * avgDist / 30));
+        SmartDashboard.putString("vision/" + m_cameraName + "/standardDeviation-state", "good :)");
+      }
+      SmartDashboard.putNumber("vision/" + m_cameraName + "/standardDeviation-average-distance", avgDist);
+    }
+  }
+
+  public boolean isEstPoseJumpy() {
+    if (m_lastEstPoses.size() < VisionConstants.NUM_LAST_EST_POSES) {
+      return true;
+    }
+
+    double totalDistance = 0;
+
+    for (int i = 0; i < m_lastEstPoses.size() - 1; i++) {
+      // add distance between ith pose and i+1th pose
+      totalDistance += Math.abs(m_lastEstPoses.get(i).minus(m_lastEstPoses.get(i + 1)).getTranslation().getNorm());
+    }
+
+    double avgDist = totalDistance / m_lastEstPoses.size();
```

Comment:
We should probably divide by the time between the first and last reading so that we get average speed instead of average time, that way the behavior is consistent no matter how fast the vision readings come in (for example, if we change resolution). 

Be sure to sanity check the time delta.